examples/sagemaker-pipelines-graphbolt/convert_ogb_papers100m_to_gconstruct.py

```diff
@@ -0,0 +1,307 @@
+    node_year = np.load(input_dir / "raw" / "node_year.npy", mmap_mode="r")
+    edge_index = np.load(input_dir / "raw" / "edge_index.npy", mmap_mode="r")
+    labels = np.load(input_dir / "raw" / "node-label.npz", mmap_mode="r")["node_label"]
+
+    num_nodes, num_features = node_feat.shape
+    num_edges = edge_index.shape[1]
+    logging.info(
+        f"Node features shape: {node_feat.shape}, Number of edges: {num_edges:,}"
+    )
+
+    # Define schemas for nodes and edges
+    node_schema = pa.schema(
+        [
+            ("nid", pa.int64()),
+            ("feat", pa.large_list(pa.float32())),
+            ("label", pa.float32()),
+            ("year", pa.int16()),
+        ]
+    )
+    edge_schema = pa.schema([("src", pa.int64()), ("dst", pa.int64())])
+
+    # Calculate chunk sizes and number of workers based on available memory
+    available_ram = psutil.virtual_memory().available
+
+    # Calculate memory usage per node row
+    node_row_bytes = (
+        num_features * 4 + 8 + 2
+    )  # 4 bytes per float32, 8 bytes for int64 nid, 2 bytes for int16 year
+    # Set node chunk size to fit within 1GB or the total number of nodes, whichever is smaller
+    node_chunk_size = min((1024**3) // node_row_bytes, num_nodes)
+
+    # Calculate memory usage per edge row
+    edge_row_bytes = 16  # 8 bytes for each int64 (src and dst)
+    # Set edge chunk size to fit within 1GB or the total number of edges, whichever is smaller
+    edge_chunk_size = min((1024**3) // edge_row_bytes, num_edges)
+
+    # Set the number of worker threads
+    # Use 2 times the number of CPU cores (or 4 if CPU count can't be determined)
+    # But limit based on available RAM, assuming each worker might use up to 2GB
+    max_workers = min(16, available_ram // (2 * 1024**3))
+
+    logging.info(
+        f"Node chunk size: {node_chunk_size:,} rows, Edge chunk size: {edge_chunk_size:,} rows."
+    )
+    logging.info(f"Max concurrent workers: {max_workers}")
+
+    with ProcessPoolExecutor(max_workers=max_workers) as executor:
+        futures = []
+
+        num_node_chunks = num_nodes // node_chunk_size
+        # Process and upload nodes in chunks
+        for idx, start in enumerate(
+            tqdm(range(0, num_nodes, node_chunk_size)), start=1
+        ):
+            end = min(start + node_chunk_size, num_nodes)
+            logging.info(
+                f"Reading data chunk {idx}/{num_node_chunks} for nodes {start:,}-{end:,}"
+            )
+            data = [
+                pa.array(range(start, end)),
+                pa.array(list(node_feat[start:end])),
+                pa.array(labels[start:end].squeeze()),
+                pa.array(node_year[start:end].astype(np.int16).squeeze()),
+            ]
+            logging.info(f"Submitting job {idx} for nodes {start:,}-{end:,}")
+            futures.append(
+                executor.submit(
+                    process_and_upload_chunk,
+                    data,
```

Comment:
Will this be very slow as you are pickling/unpickling the entire dataset.

Reply:
Yeah the time gains are not crazy here, but the pickling seem to quite efficient. I tried using threads to see if we can get some concurrency benefit but that was actually slower that using processes.